### ISSUE TITLE
horovod SHUT_DOWN_ERROR design issue

### ISSUE THREAD

**Issue by LiweiPeng** [horovod/horovod#694]
In current horovod design, the SHUT_DOWN_ERROR is converted to tensorflow unknown error. However, this has some issues when I use tf estimator for training:
1) estimator.train() won't save the last step before Unknown error occurs. This results loss of training steps.
2) estimator printed long stack trace which is difficult to follow.

An expected design is to throw python StopIteration exception. tf.estimator handles StopIteration and will stop training gracefully.

Another solution is to throw tf OutOfRange error. The bad part is that SHUT_DOWN_ERROR is not out of range.

**Comment by tgaddair** [comment]
Hey @LiweiPeng, would this change be needed only for the SHUT_DOWN_ERROR or for any instance where we raise an UnknownError?

**Comment by LiweiPeng** [comment]
@tgaddair  This change is desired for any case where horovod cannot proceed normally and has to stop/abort. SHUT_DOWN_ERROR is such a case.

Horovod also throws other UnknownErrors. But I am not sure whether these errors will result in horovod to be stopped or aborted. If they do, it is desired that they are handled in the same way.

**Comment by tgaddair** [comment]
@LiweiPeng Some of the error types TensorFlow supports can cause issues as was the case with #552. Would CancelledError work in your case?  And do you have a simple repro we can use to verify the change works?

**Comment by alsrgv** [comment]
@LiweiPeng, can you share more light on your use case?  Do you get this error because you have an uneven number of steps across workers, or some workers actually physically fail?  If you're working around failure, is it possible to checkpoint more frequently to avoid losing much of progress?

The OutOfRange prototype we have in #711 currently only works with Estimator, while our API is designed to work across all APIs.  It's also a bit strange to throw OutOfRange error when the real error is IllegalState (cluster is shut down / error happened).

**Comment by abditag2** [comment]
https://github.com/uber/horovod/pull/711

**Comment by abditag2** [comment]
@LiweiPeng Happy new year! Can you look at @alsrgv comments?

**Comment by LiweiPeng** [comment]
@abditag2 My use case is exactly what's inside  #711. The implementation is what I want to have for Horovod.

**Comment by alsrgv** [comment]
@LiweiPeng, can you share more light on your use case? Do you get this error because you have an uneven number of steps across workers, or some workers actually physically fail? If you're working around failure, is it possible to checkpoint more frequently to avoid losing much of progress?

The OutOfRange prototype we have in #711 currently only works with Estimator, while our API is designed to work across all APIs. It's also a bit strange to throw OutOfRange error when the real error is IllegalState (cluster is shut down / error happened).

**Comment by LiweiPeng** [comment]
@alsrgv I am using Estimator. My use case is that I have an uneven number of steps across workers. There is no worker that is physically failed. 

The uneven number of steps is hard to get around formy NMT application. When I was given a new dataset for training on N gpus, I couldn't find a way to really evenly distributed the batches among the N gpus. So it often results that some GPUs has 1 or 2 batches than others. And the uneven distribution can result in shutdown error.

**Comment by alsrgv** [comment]
@LiweiPeng, my concern with #711 is that it's specific to Estimator, but requires a change in core Horovod API, and also it has the potential to hide actual failures - e.g. if you have an exception on one of the workers, your training will shut down with what may appear as successful completion.

How are you sharding your data?  Would using [TensorFlow dataset API](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#shard) help you?  Alternatively, you could pre-compute approximate # of steps per epoch, and broadcast it to all the workers in the beginning.

**Comment by LiweiPeng** [comment]
@alsrgv I understand this is specific to Estimator. One implementation option is to define an environment variable to switch this feature on/off, with  default to be current existing design. 

I shard my data using tensorflow shard.  I have to shard in number of lines (tf.data.TextLineDataset) because NMT handles lines. the results are not even when in terms of words at training. so it often hits this shutdown error.

**Comment by alsrgv** [comment]
@LiweiPeng, I feel like solving this via shutdown error is a big hack.  Are you training based on the number of epochs?  Could you switch to training based on # of total batches (that you'd divide by # of workers)?  That way you know precisely how many N-grams your training did process.

We can expose additional API to notify Horovod about end-of-data situation on one of the ranks that you could call at the end of your training, which would throw OutOfRange on other ranks -- but I feel like this could also be resolved by using an explicit number of batches.

**Comment by LiweiPeng** [comment]
@alsrgv I am training based on number of epochs. Let me change it to using total batches. I found that I could leverage the tf.data.Dataset.repeat() API to avoid the issue.

**Comment by alsrgv** [comment]
@LiweiPeng, perfect. That's exactly what I had in mind.

**Comment by LiweiPeng** [comment]
I validated the above solution and it works perfect for me. I am closing this issue.

**Comment by alsrgv** [comment]
@LiweiPeng, FYI - we started another discussion in #832 around more general-purpose solution for uneven steps.  Can you see if it would be helpful for your use cases as well?

**Comment by LiweiPeng** [comment]
Thanks Alex. I'll follow up on the other thread.
